tests/base/test_pytorch_transforms.py

```diff
@@ -19,24 +32,35 @@ def test_module_mapping_transform():
     class TestTransform(ModuleMappingTransform):
         _module_mapping = {nn.Linear: nn.Identity}
 
-    class TestModel(nn.Module):
-        def __init__(self):
-            super().__init__()
+    model = TestModel()
+    x = torch.rand(1, 32)
+    y1 = model(x)
+    assert torch.any(y1 != x)
+
+    model, transformed = TestTransform.apply(model)
+    assert transformed
+    y2 = model(x)
+    assert torch.all(y2 == x)
 
-            self.a = nn.Linear(32, 64)
-            self.b = nn.Linear(64, 32)
 
-        def forward(self, x):
-            x = self.a(x)
-            x = self.b(x)
-            return x
+def test_module_mutator_transform():
+    with pytest.raises(TypeError):
+        ModuleMutatorTransform()
+
+    class TestTransform(ModuleMutatorTransform):
+        _match_class = nn.Linear
+
+        @classmethod
+        def mutate(cls, original_module: nn.Module, parent_module: nn.Module):
+            return nn.Identity()
```

Comment:
In this test, we can easily see that the functionality of ModuleMutatorTransform and ModuleMappingTransform are basically the same.

Reply:
They are not same, check line 64 of this file.
`ModuleMapping` is when you don't need to initialize new class, and it is possible to directly assign `old_module.__class__=some_new_class`
`ModuleMutator` is when you must initialize a new class and replace in place of old one. Only class mapping is not enough. In AWQ/GPTQ we need to initialize a new class and then pack the unpacked weights into new class object.

Reply:
I did some thinking on using `__qeff_init__`.

- Approach 1: Using `__qeff_init__`  in `WQLinearGEMM` class
    - Moving unpacking, initialization and repacking to `__qeff_init__`
        If we do this and return the new `QuantLinearORT` class as output, we still need a way to insert this class into model i.e. `setattr` call -> which is happening in `ModuleMutator.apply` method. In this case we are moving away from what `__qeff_init__` is supposed to do, it needs to change the module in-place without external help. And also, In this case `WQLinearGEMM` has the code of packing for `QuantLinearORT` which it should not care about.
- Apprach 2:  Using `__qeff_init__` in Attention clasess
   - This leads to us adding `__qeff_init__` in all models Attention classes i.e. separate for `LlamaAttention`, `MistralAttention` and so on. Not scalable.

I think `ModuleMutator` can be quiete useful when we want to

Reply:
If developer doesn't apply `KVCacheTransform` but applies `AWQToMatmulNbitsTransform` the model will give wrong output as weights are not packed yet, and we can't force user to always run `KVCacheTransform`

Reply:
I'm not sure how this has anything to do with `KVCacheTransform`. If you see closely, we have to add the `__qeff_init__` in the `QuantLinearORT`, not in the KVCache classes. `KVCacheTransform` and `AWQToMatmulNbitsTransform` will be completely orthogonal.

Reply:
- We set `__class__` exactly to change the type of a variable at runtime based on runtime conditions. Here the runtime condition is that the user wants to run the pytorch module on QAIC and not on other devices available in pytorch.
- If we initialize a new class and then do `setattr`, we'll be duplicating the weights of the model in memory. The primary reason we use `a.__class__ = B` is to not duplicate the weights unnecessarily, we only change the behaviour of `forward` in the class.
- Again, if we want to switch to `setattr` instead of `a.__class__ = B` everywhere, we have to know parameters to be passed to `B()` initialization, which may be different for each class. Creating a new ModuleMutator for each and every class that we need to replace is not scalable.
- Using `ModuleMutator` actually scatters the related code across in multiple files. For example, here part of the implementation is in Transform and part of it is in the QuantLinearORT class.
- Rotary embedding implementation is different for different models, some use default embeddings, some use scaled embeddings. There is also rotary_dim config, which will change the way embeddings are applied. So, we can't really re-use the implementation.

ModuleMutator is almost the same as ModuleMapping:
- You could simple move the `mutate` function to `__qeff_init__` in the target class.
- This will keep the code for that specific class in there, instead of having it in a different file (transforms).
- In AWQ case, we have to make changes in weights and drop old weights. We could make all the changes and then delete the old weights simply with `del self.old_weight`

Issue with ModuleMutator design:
- You can match only one class; this works well for AWQ currently. Say tomorrow there are new modules to support conv and norm as WQConvGEMM and WQLayerNormGEMM. Now you have to create 2 more transforms just to target those classes.
- If you think about a solution for the issue, you'll naturally come to adding a special function (`__qeff_init__`) in the replacement class itself.

With all that said, you can continue with ModuleMutator class if you still think it's a good idea. I'd like to close this thread as it has gone longer that I expected.